mm/kfence/static_branch.c

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/atomic.h>
+#include <linux/mm.h>
+#include <linux/static_key.h>
+#include <linux/wait.h>
+#include <linux/workqueue.h>
+
+#include "kfence.h"
+#include "../slab.h"
+
+/* The static key to set up a KFENCE allocation. */
+DEFINE_STATIC_KEY_FALSE(kfence_allocation_key);
+
+/* Gates the allocation, ensuring only one succeeds in a given period. */
+static atomic_t allocation_gate = ATOMIC_INIT(1);
+/* Wait queue to wake up heartbeat timer task. */
+static DECLARE_WAIT_QUEUE_HEAD(allocation_wait);
+
+void *kfence_alloc_with_size(struct kmem_cache *s, size_t size, gfp_t flags)
+{
+	void *ret;
+
+	if (atomic_inc_return(&allocation_gate) > 1)
+		return NULL;
+	wake_up(&allocation_wait);
+
+	if (!kfence_is_enabled())
+		return NULL;
+
+	if ((size > PAGE_SIZE) || (s->size > PAGE_SIZE))
```

Comment:
`size > PAGE_SIZE` part is excessive? Can size be larger than s->size?

Reply:
@ramosian-glider what was the reason for this?

I've added a TODO.

Reply:
`size` is either zero or something between `s->size` or `previous_malloc_cache->size`. So, yes, this is excessive.

Reply:
Thanks for clarifying. I already added a TODO, and this will all go away I guess in the eventual cleanup to merge it all back into core.c.